File: whatsapp-sticker-uploader/app/src/main/java/com/stickeruploader/MainActivity.java
package com.stickeruploader;

import android.app.Activity;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.os.Bundle;
import android.os.Handler;
import android.os.Looper;
import android.util.Log;
import android.view.View;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;
import android.widget.Button;
import android.widget.ListView;
import android.widget.TextView;
import android.widget.Toast;

import com.stickeruploader.models.StickerPack;

import java.io.File;
import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.Executor;
import java.util.concurrent.Executors;

public class MainActivity extends Activity {

    private static final String TAG = "MainActivity";

    private ListView listView;
    private TextView tvStatus;
    private Button btnReload;
    private Button btnAddAll;

    private List<StickerPack> allPacks = new ArrayList<>();
    private Executor executor = Executors.newSingleThreadExecutor();
    private Handler mainHandler = new Handler(Looper.getMainLooper());

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        Log.d(TAG, "onCreate() chiamato");

        setContentView(R.layout.activity_main);

        listView = findViewById(R.id.list);
        tvStatus = findViewById(R.id.tvStatus);
        btnReload = findViewById(R.id.btnReload);
        btnAddAll = findViewById(R.id.btnAddAll);

        setupButtons();
        checkWhatsAppInstalled();
        loadStickers();
    }

    private void setupButtons() {
        Log.d(TAG, "setupButtons()");

        btnReload.setOnClickListener(v -> {
            Log.d(TAG, "btnReload cliccato");
            loadStickers();
        });

        btnAddAll.setOnClickListener(v -> {
            Log.d(TAG, "btnAddAll cliccato - aggiungendo " + allPacks.size() + " pack");
            for (StickerPack pack : allPacks) {
                addPackToWhatsApp(pack);
            }
        });

        listView.setOnItemClickListener((parent, view, position, id) -> {
            if (position < allPacks.size()) {
                StickerPack pack = allPacks.get(position);
                Log.d(TAG, "Pack cliccato: " + pack.identifier);
                addPackToWhatsApp(pack);
            }
        });
    }

    private void checkWhatsAppInstalled() {
        Log.d(TAG, "checkWhatsAppInstalled()");
        boolean installed = isPackageInstalled("com.whatsapp") || isPackageInstalled("com.whatsapp.w4b");
        Log.d(TAG, "WhatsApp installato: " + installed);

        if (!installed) {
            showError("WhatsApp non è installato");
        }
    }

    private boolean isPackageInstalled(String packageName) {
        try {
            getPackageManager().getPackageInfo(packageName, 0);
            return true;
        } catch (PackageManager.NameNotFoundException e) {
            return false;
        }
    }

    private void loadStickers() {
        Log.d(TAG, "loadStickers() iniziato");
        tvStatus.setText("Caricamento sticker...");
        btnReload.setEnabled(false);
        btnAddAll.setEnabled(false);

        executor.execute(() -> {
            try {
                Log.d(TAG, "Caricando pack da " + StickerPackLoader.STICKER_DIR.getAbsolutePath());
                List<StickerPack> packs = StickerPackLoader.loadAllPacks();
                Log.d(TAG, "Caricati " + packs.size() + " pack");

                allPacks = packs;

                mainHandler.post(() -> {
                    if (packs.isEmpty()) {
                        Log.w(TAG, "Nessun pack trovato");
                        showError("Nessuno sticker trovato");
                    } else {
                        updateUI(packs);
                    }
                });
            } catch (Exception e) {
                Log.e(TAG, "Errore caricamento", e);
                mainHandler.post(() -> showError("Errore: " + e.getMessage()));
            }
        });
    }

    private void updateUI(List<StickerPack> packs) {
        Log.d(TAG, "updateUI() con " + packs.size() + " pack");
        tvStatus.setText(packs.size() + " pack trovati");
        btnReload.setEnabled(true);
        btnAddAll.setEnabled(true);

        List<String> packNames = new ArrayList<>();
        for (StickerPack pack : packs) {
            packNames.add(pack.name + " (" + pack.stickers.size() + " sticker)");
        }

        ArrayAdapter<String> adapter = new ArrayAdapter<>(this,
                android.R.layout.simple_list_item_1, packNames);
        listView.setAdapter(adapter);
    }

    private void addPackToWhatsApp(StickerPack pack) {
        Log.d(TAG, "addPackToWhatsApp() per " + pack.identifier);
        tvStatus.setText("Copiamento file di '" + pack.name + "'...");

        executor.execute(() -> {
            try {
                Log.d(TAG, "Copiando file del pack " + pack.identifier);
                copyPackFilesToAppStorage(pack);

                Log.d(TAG, "File copiati, inviando intent a WhatsApp");
                mainHandler.post(() -> sendWhatsAppIntent(pack));
            } catch (Exception e) {
                Log.e(TAG, "Errore copia file", e);
                mainHandler.post(() -> showError("Errore: " + e.getMessage()));
            }
        });
    }

    private void copyPackFilesToAppStorage(StickerPack pack) {
        Log.d(TAG, "copyPackFilesToAppStorage() per " + pack.identifier);

        File stickersDir = new File(getFilesDir(), "stickers");
        File trayDir = new File(getFilesDir(), "tray_images");

        if (!stickersDir.exists()) {
            stickersDir.mkdirs();
            Log.d(TAG, "Directory creata: " + stickersDir.getAbsolutePath());
        }
        if (!trayDir.exists()) {
            trayDir.mkdirs();
            Log.d(TAG, "Directory creata: " + trayDir.getAbsolutePath());
        }

        // Copia sticker files
        for (int i = 0; i < pack.stickers.size(); i++) {
            com.stickeruploader.models.Sticker sticker = pack.stickers.get(i);
            File sourceFile = new File(StickerPackLoader.STICKER_DIR, sticker.imageFileName);
            File destFile = new File(stickersDir, sticker.imageFileName);

            if (!destFile.exists() && sourceFile.exists()) {
                Log.d(TAG, "Copiando sticker: " + sticker.imageFileName);
                try {
                    sourceFile.renameTo(destFile) || copyFile(sourceFile, destFile);
                    Log.d(TAG, "Copiato: " + destFile.getAbsolutePath());
                } catch (Exception e) {
                    Log.e(TAG, "Errore copia file: " + e.getMessage());
                }
            }
        }

        // Copia tray image
        File sourceTrayFile = new File(StickerPackLoader.STICKER_DIR, pack.trayImageFile);
        File destTrayFile = new File(trayDir, pack.identifier + "_tray.webp");

        if (!destTrayFile.exists() && sourceTrayFile.exists()) {
            Log.d(TAG, "Copiando tray image: " + pack.trayImageFile);
            try {
                sourceTrayFile.renameTo(destTrayFile) || copyFile(sourceTrayFile, destTrayFile);
                Log.d(TAG, "Copiato tray: " + destTrayFile.getAbsolutePath());
            } catch (Exception e) {
                Log.e(TAG, "Errore copia tray: " + e.getMessage());
            }
        }

        Log.d(TAG, "copyPackFilesToAppStorage() completato per " + pack.identifier);
    }

    private boolean copyFile(File source, File dest) throws Exception {
        try (java.io.FileInputStream input = new java.io.FileInputStream(source);
             java.io.FileOutputStream output = new java.io.FileOutputStream(dest)) {
            byte[] buffer = new byte[8192];
            int length;
            while ((length = input.read(buffer)) > 0) {
                output.write(buffer, 0, length);
            }
            return true;
        }
    }

    private void sendWhatsAppIntent(StickerPack pack) {
        Log.d(TAG, "sendWhatsAppIntent() per " + pack.identifier);

        Intent intent = new Intent();
        intent.setAction("com.whatsapp.intent.action.ENABLE_STICKER_PACK");
        intent.putExtra("sticker_pack_id", pack.identifier);
        intent.putExtra("sticker_pack_authority", StickerContentProvider.AUTHORITY);
        intent.putExtra("sticker_pack_name", pack.name);
        intent.putExtra("sticker_pack_publisher", pack.publisher);
        intent.putExtra("sticker_pack_animated_emojis", false);
        intent.setPackage("com.whatsapp");

        Log.d(TAG, "Intent: " + intent.getAction());
        Log.d(TAG, "  pack_id=" + pack.identifier);
        Log.d(TAG, "  authority=" + StickerContentProvider.AUTHORITY);

        try {
            startActivity(intent);
            Log.d(TAG, "Intent inviato con successo");
            tvStatus.setText("Intent inviato a WhatsApp");
        } catch (Exception e) {
            Log.e(TAG, "Errore invio intent", e);
            showError("Errore: " + e.getMessage());
        }
    }

    private void showError(String message) {
        Log.e(TAG, "Errore: " + message);
        Toast.makeText(this, message, Toast.LENGTH_LONG).show();
        tvStatus.setText("Errore: " + message);
    }
}

File: whatsapp-sticker-uploader/app/src/main/java/com/stickeruploader/MainActivity.kt
package com.stickeruploader

import android.Manifest
import android.content.Intent
import android.content.pm.PackageManager
import android.net.Uri
import android.os.Build
import android.os.Bundle
import android.os.Environment
import android.provider.Settings
import android.view.View
import android.widget.Toast
import androidx.activity.result.contract.ActivityResultContracts
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatActivity
import androidx.core.content.ContextCompat
import androidx.recyclerview.widget.LinearLayoutManager
import android.widget.SeekBar
import com.stickeruploader.databinding.ActivityMainBinding
import com.stickeruploader.models.StickerPack

class MainActivity : AppCompatActivity() {

    private companion object {
        const val TAG = "MainActivity"
    }

    private lateinit var binding: ActivityMainBinding
    private lateinit var adapter: StickerPackAdapter
    private var stickersPerPack = StickerPackLoader.STICKERS_PER_PACK
    private var waitingForManageStoragePermission = false

    private val ADD_PACK_REQUEST_CODE = 200

    private val permissionLauncher = registerForActivityResult(
        ActivityResultContracts.RequestPermission()
    ) { granted ->
        if (granted) loadStickers() else showPermissionDeniedDialog()
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityMainBinding.inflate(layoutInflater)
        setContentView(binding.root)

        AppLogger.init(this)
        AppLogger.separator("APP AVVIATA")
        AppLogger.i(TAG, "Android ${Build.VERSION.RELEASE} API ${Build.VERSION.SDK_INT}")

        binding.tvStatus.text = "Log: ${AppLogger.getLogFilePath()}"

        setupSlider()
        setupRecyclerView()
        setupButtons()
        checkPermissionsAndLoad()
    }

    override fun onResume() {
        super.onResume()
        if (waitingForManageStoragePermission &&
            Build.VERSION.SDK_INT >= Build.VERSION_CODES.R &&
            Environment.isExternalStorageManager()) {
            waitingForManageStoragePermission = false
            loadStickers()
        }
    }

    private fun setupSlider() {
        // SeekBar: progress 0..27 → valore reale 3..30
        binding.seekBarPackSize.progress = stickersPerPack - 3
        binding.tvPackSize.text = "Sticker per pacchetto: $stickersPerPack"

        binding.seekBarPackSize.setOnSeekBarChangeListener(object : SeekBar.OnSeekBarChangeListener {
            override fun onProgressChanged(seekBar: SeekBar, progress: Int, fromUser: Boolean) {
                stickersPerPack = progress + 3
                binding.tvPackSize.text = "Sticker per pacchetto: $stickersPerPack"
            }
            override fun onStartTrackingTouch(seekBar: SeekBar) {}
            override fun onStopTrackingTouch(seekBar: SeekBar) {
                loadStickers()
            }
        })
    }

    private fun setupRecyclerView() {
        adapter = StickerPackAdapter(onAddClick = { pack -> addPackToWhatsApp(pack) })
        binding.recyclerView.layoutManager = LinearLayoutManager(this)
        binding.recyclerView.adapter = adapter
    }

    private fun setupButtons() {
        binding.btnAddAll.setOnClickListener {
            val remaining = adapter.getPacks().filter { !it.isAddedToWhatsApp }
            if (remaining.isEmpty()) {
                Toast.makeText(this, "Tutti i pack sono già stati aggiunti!", Toast.LENGTH_SHORT).show()
                return@setOnClickListener
            }
            AlertDialog.Builder(this)
                .setTitle("Aggiungi tutti i pack")
                .setMessage("Stai per aggiungere ${remaining.size} pack a WhatsApp.\n\nPer ogni pack apparirà una finestra di conferma WhatsApp.\n\nVuoi continuare?")
                .setPositiveButton("Sì, inizia") { _, _ -> addAllPacksSequentially(remaining, 0) }
                .setNegativeButton("Annulla", null)
                .show()
        }

        binding.btnReload.setOnClickListener { checkPermissionsAndLoad() }

        binding.btnShareLog.setOnClickListener { shareLog() }
    }

    private fun shareLog() {
        val path = AppLogger.getLogFilePath()
        val file = java.io.File(path)
        if (!file.exists()) {
            Toast.makeText(this, "File di log non trovato", Toast.LENGTH_SHORT).show()
            return
        }
        val text = try { file.readText() } catch (e: Exception) {
            Toast.makeText(this, "Errore lettura log: ${e.message}", Toast.LENGTH_LONG).show()
            return
        }
        val intent = Intent(Intent.ACTION_SEND).apply {
            type = "text/plain"
            putExtra(Intent.EXTRA_SUBJECT, "StickerUploader Log")
            putExtra(Intent.EXTRA_TEXT, text)
        }
        startActivity(Intent.createChooser(intent, "Condividi log"))
    }

    private fun checkPermissionsAndLoad() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {
            if (!Environment.isExternalStorageManager()) {
                showManageStorageDialog(); return
            }
        } else {
            if (ContextCompat.checkSelfPermission(this, Manifest.permission.READ_EXTERNAL_STORAGE)
                != PackageManager.PERMISSION_GRANTED) {
                permissionLauncher.launch(Manifest.permission.READ_EXTERNAL_STORAGE); return
            }
        }
        loadStickers()
    }

    private fun showManageStorageDialog() {
        AlertDialog.Builder(this)
            .setTitle("Accesso ai file richiesto")
            .setMessage("Per leggere gli sticker WhatsApp da:\n\n${StickerPackLoader.STICKER_DIR.absolutePath}\n\nserve il permesso 'Accesso a tutti i file'.\n\nTocca OK → attiva l'interruttore per Sticker Uploader → torna all'app.")
            .setPositiveButton("Apri Impostazioni") { _, _ ->
                waitingForManageStoragePermission = true
                try {
                    startActivity(Intent(Settings.ACTION_MANAGE_APP_ALL_FILES_ACCESS_PERMISSION).apply {
                        data = Uri.fromParts("package", packageName, null)
                    })
                } catch (e: Exception) {
                    startActivity(Intent(Settings.ACTION_MANAGE_ALL_FILES_ACCESS_PERMISSION))
                }
            }
            .setNegativeButton("Annulla") { _, _ ->
                binding.tvStatus.text = "Permesso necessario. Premi 'Ricarica' dopo averlo concesso."
            }
            .show()
    }

    private fun loadStickers() {
        binding.progressBar.visibility = View.VISIBLE
        binding.tvStatus.text = "Validazione sticker animati in corso..."
        binding.recyclerView.visibility = View.GONE

        Thread {
            AppLogger.separator("CARICAMENTO STICKER ANIMATI (${stickersPerPack} per pack)")
            val packs = StickerPackLoader.loadAllPacks(stickersPerPack)
            val validCount = StickerPackLoader.lastValidCount
            val invalidCount = StickerPackLoader.lastInvalidCount
            AppLogger.i(TAG, "Pack animati caricati: ${packs.size} (✅$validCount sticker validi, ❌$invalidCount scartati)")

            val items = mutableListOf<PackItem>()
            if (packs.isNotEmpty()) {
                items.add(PackItem.Header("🎬 Animati — ${packs.size} pack, ${packs.sumOf { it.stickers.size }} sticker"))
                packs.forEach { items.add(PackItem.Pack(it)) }
            }

            runOnUiThread {
                binding.progressBar.visibility = View.GONE
                adapter.updateItems(items)
                binding.recyclerView.visibility = View.VISIBLE

                if (packs.isEmpty()) {
                    binding.tvStatus.text = "Nessuno sticker animato valido trovato in:\n${StickerPackLoader.STICKER_DIR.absolutePath}\n\n✅ Validi: $validCount · ❌ Scartati (loop≠0 o dim≠512px): $invalidCount"
                    binding.btnAddAll.isEnabled = false
                } else {
                    val total = packs.sumOf { it.stickers.size }
                    binding.tvStatus.text = "${packs.size} pack animati · $total sticker · ✅$validCount validi · ❌$invalidCount scartati"
                    binding.btnAddAll.isEnabled = true
                }
            }
        }.start()
    }

    private fun addPackToWhatsApp(pack: StickerPack) {
        binding.progressBar.visibility = View.VISIBLE
        binding.tvStatus.text = "Preparazione ${pack.name}..."

        Thread {
            AppLogger.separator("AGGIUNTA PACK: ${pack.identifier}")
            try {
                StickerFileCache.preparePack(applicationContext, pack)
            } catch (e: Exception) {
                AppLogger.e(TAG, "Errore preparazione pack", e)
            }

            runOnUiThread {
                binding.progressBar.visibility = View.GONE
                val intent = Intent().apply {
                    action = "com.whatsapp.intent.action.ENABLE_STICKER_PACK"
                    putExtra("sticker_pack_id", pack.identifier)
                    putExtra("sticker_pack_authority", StickerContentProvider.AUTHORITY)
                    putExtra("sticker_pack_name", pack.name)
                }
                try {
                    @Suppress("DEPRECATION")
                    startActivityForResult(intent, ADD_PACK_REQUEST_CODE)
                    pendingPackId = pack.identifier
                } catch (e: Exception) {
                    AppLogger.e(TAG, "Errore invio intent", e)
                    Toast.makeText(this, "Errore: ${e.message}", Toast.LENGTH_LONG).show()
                }
            }
        }.start()
    }

    private var pendingPackId: String? = null
    private var batchPacksToAdd: List<StickerPack> = emptyList()
    private var batchCurrentIndex: Int = 0

    private fun addAllPacksSequentially(packs: List<StickerPack>, startIndex: Int) {
        if (startIndex >= packs.size) {
            Toast.makeText(this, "Tutti i pack sono stati aggiunti!", Toast.LENGTH_SHORT).show()
            return
        }
        batchPacksToAdd = packs
        batchCurrentIndex = startIndex
        addPackToWhatsApp(packs[startIndex])
    }

    @Suppress("DEPRECATION")
    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        if (requestCode != ADD_PACK_REQUEST_CODE) return

        val packId = pendingPackId
        AppLogger.i(TAG, "Risultato WhatsApp: resultCode=$resultCode, packId=$packId")

        if (resultCode == RESULT_OK && packId != null) {
            AppLogger.i(TAG, "✅ Pack aggiunto")
            adapter.markAsAdded(packId)
            if (batchPacksToAdd.isNotEmpty()) {
                batchCurrentIndex++
                if (batchCurrentIndex < batchPacksToAdd.size) {
                    binding.recyclerView.postDelayed({
                        addAllPacksSequentially(batchPacksToAdd, batchCurrentIndex)
                    }, 500)
                } else {
                    batchPacksToAdd = emptyList()
                    Toast.makeText(this, "Tutti i pack aggiunti!", Toast.LENGTH_SHORT).show()
                }
            }
        } else {
            AppLogger.w(TAG, "❌ Pack NON aggiunto - resultCode=$resultCode")
            data?.extras?.keySet()?.forEach { key ->
                AppLogger.w(TAG, "  extra[$key] = ${data.extras?.get(key)}")
            }
            if (resultCode == RESULT_CANCELED) {
                Toast.makeText(this, "Aggiunta annullata", Toast.LENGTH_SHORT).show()
                batchPacksToAdd = emptyList()
            }
        }
        pendingPackId = null
    }

    private fun showPermissionDeniedDialog() {
        AlertDialog.Builder(this)
            .setTitle("Permesso necessario")
            .setMessage("L'app ha bisogno del permesso storage per leggere gli sticker.")
            .setPositiveButton("Riprova") { _, _ -> checkPermissionsAndLoad() }
            .setNegativeButton("Annulla") { _, _ -> binding.tvStatus.text = "Permesso negato." }
            .show()
    }
}
